x-pack/plugins/security/server/saved_objects/audit_events.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { AuditEventDecorator } from '../../../../../src/core/server';
+
+export interface SavedObjectEventArgs {
+  action: string;
+  objects: ReadonlyArray<{
+    type: string;
+    id?: string;
+    namespaces?: string[];
+  }>;
+  error?: Error;
+}
+
+export const savedObjectCreateEvent: AuditEventDecorator<SavedObjectEventArgs> = (
+  event,
+  { action, objects, error }
+) => {
+  const doc =
+    objects.length === 1
+      ? `saved object '${objects[0].id}' of type '${objects[0].type}'`
+      : `${objects.length} saved objects`;
+
+  return {
+    ...event,
+    message: error
+      ? `Failed attempt to create ${doc} by user '${event.user.name}'`
+      : `User '${event.user.name}' created ${doc}`,
+    event: {
+      action,
+      category: 'database',
+      type: 'creation',
+      outcome: error ? 'failure' : 'success',
+    },
+    error: error
+      ? {
+          code: error.name,
+          message: error.message,
+        }
+      : undefined,
+    kibana: {
+      ...event.kibana,
+      saved_objects: objects.map(({ type, id }) => ({ type, id })),
+    },
+  };
+};
+
+export const savedObjectReadEvent: AuditEventDecorator<SavedObjectEventArgs> = (
+  event,
+  { action, objects, error }
+) => {
+  const doc =
+    objects.length === 1
+      ? `saved object '${objects[0].id}' of type '${objects[0].type}'`
+      : `${objects.length} saved objects`;
+
+  return {
+    ...event,
+    message: error
+      ? `Failed attempt to access ${doc} by user '${event.user.name}'`
+      : `User '${event.user.name}' accessed ${doc}`,
+    event: {
+      action,
+      category: 'database',
+      type: 'access',
+      outcome: error ? 'failure' : 'success',
+    },
+    error: error
+      ? {
+          code: error.name,
+          message: error.message,
+        }
+      : undefined,
+    kibana: {
+      ...event.kibana,
+      saved_objects: objects.map(({ type, id }) => ({ type, id })),
+    },
+  };
+};
+
+export const savedObjectUpdateEvent: AuditEventDecorator<SavedObjectEventArgs> = (
+  event,
+  { action, objects, error }
+) => {
+  const doc =
+    objects.length === 1
+      ? `saved object '${objects[0].id}' of type '${objects[0].type}'`
+      : `${objects.length} saved objects`;
```

Comment:
I couldn't find an example of this. Can you point me to one?

Reply:
Sorry for chiming in :see_no_evil: I believe Joe meant [getDefaultTtile](https://github.com/elastic/kibana/blob/358d13919bc597d13e6a98ec97fae2615093027e/src/plugins/saved_objects_management/public/lib/get_default_title.ts#L20) we use in a number of places.

Reply:
It's closer to what I originally intended but I thought it would look a bit strange in a sentence for saved object types that aren't a single word (e.g. `ingest_manager_settings`, `canvas-workpad-template`)

- User 'elastic' accessed canvas-workpad-template [id=029bdeb3-40a6-4c90-9320-a5566abaf427]
- User 'elastic' accessed saved object '029bdeb3-40a6-4c90-9320-a5566abaf427' of type 'canvas-workpad-template'